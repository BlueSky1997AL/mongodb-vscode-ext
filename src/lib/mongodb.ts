'use strict';

import MongoDBMemoryServer from 'mongodb-memory-server';
import { homedir } from 'os';
import { existsSync, mkdirsSync } from 'fs-extra';
import * as vscode from 'vscode';

import statusBarItem from './status-bar-item';

class MongoDB {

  public mongoDBInstance = null;

  start(options: any = {}) {
    const { port, dbName, dbPath, storageEngine, instanceDebug, binaryVersion, downloadDir, platform, arch, binaryDebug, mongoDebug, autoStart } = options

    let finalDbPath = dbPath || `${homedir()}/.mongodb/data`;
    if (!existsSync(finalDbPath)) {
      mkdirsSync(finalDbPath);
    }

    const mongod = new MongoDBMemoryServer({
      instance: {
        port: parseInt(port) || 27017,
        dbName: dbName || 'db',
        dbPath: finalDbPath,
        storageEngine,
        debug: instanceDebug || false
      },
      binary: {
        version: binaryVersion,
        downloadDir: downloadDir || `${homedir()}/.mongodb/binaries`,
        platform,
        arch,
        debug: binaryDebug
      },
      debug: mongoDebug,
      autoStart
    });

    mongod.start();

    this.mongoDBInstance = mongod;

    statusBarItem.update({
      text: '$(database) MongoDB',
      command: 'extension.showMenu',
      color: '#69b241',
      tooltip: 'MongoDB 数据库已启动, 单击以查看选项'
    })
  }

  stop() {
    this.mongoDBInstance.stop();
    this.mongoDBInstance = null;
    statusBarItem.update({
      text: '$(database) MongoDB',
      command: 'extension.showMenu',
      tooltip: 'MongoDB 数据库未启动，单击启动'
    });
  }

  get instance() {
    return this.mongoDBInstance;
  }

}

export default new MongoDB();
