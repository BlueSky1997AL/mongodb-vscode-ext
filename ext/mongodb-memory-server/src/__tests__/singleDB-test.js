/* @flow */

import { MongoClient } from 'mongodb';
import MongoDBMemoryServer from '../MongoMemoryServer';

jasmine.DEFAULT_TIMEOUT_INTERVAL = 60000;

let con;
let db;
let mongoServer;
beforeAll(async () => {
  mongoServer = new MongoDBMemoryServer();
  const mongoUri = await mongoServer.getConnectionString();
  con = await MongoClient.connect(mongoUri);
  db = con.db(await mongoServer.getDbName());
});

afterAll(() => {
  con.close();
  mongoServer.stop();
});

describe('Single mongoServer', () => {
  it('should start mongo server', async () => {
    expect(db).toBeDefined();
    const col = db.collection('test');
    const result = await col.insert([{ a: 1 }, { b: 1 }]);
    expect(result.result).toMatchSnapshot();
    expect(await col.count({})).toBe(2);
  });
});
