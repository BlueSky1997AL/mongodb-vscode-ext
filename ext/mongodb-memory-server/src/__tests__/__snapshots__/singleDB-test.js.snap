// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single mongoServer should start mongo server 1`] = `
Object {
  "n": 2,
  "ok": 1,
}
`;
